XMonad/Layout/FixedAspectRatio.hs

```diff
@@ -0,0 +1,161 @@
+-- the 'layoutHook'.
+--
+-- You also want to add keybindings to set and clear the aspect ratio:
+--
+-- >      -- Set the aspect ratio of the focused window to 16:9
+-- >   ,((modm, xK_a), withFocused $ sendMessage . FixRatio (16 / 9))
+-- >
+-- >      -- Clear the aspect ratio from the focused window
+-- >   ,((modm .|. shiftMask, xK_a), withFocused $ sendMessage . ResetRatio)
+--
+-- There's one caveat: to keep the usage of the modifier simple, it
+-- doesn't remove a window from its cache automatically. Which means
+-- that if you close a program window that has some fixed aspect ratios
+-- and relaunch it, sometimes it'll still have the fixed aspect ratio.
+-- You can try to avoid this by changing they keybinding used to kill
+-- the window:
+--
+-- >  , ((modMask .|. shiftMask, xK_c), withFocused (sendMessage . ResetRatio) >> kill)
+--
+-- See "XMonad.Doc.Extending#Editing_key_bindings" for more info
+-- on customizing the keybindings.
+--
+-- This layout also comes with a 'ManageHook' 'doFixAspect' to
+-- automatically fix the aspect ratio:
+--
+-- > myManageHook = composeOne [
+-- >   title =? "Netflix" <||> className =? "vlc" --> doFixAspect (16 / 9)
+-- >   ...
+-- > ]
+--
+-- Check "XMonad.Doc.Extending#Editing_the_manage_hook" for more information on
+-- customizing the manage hook.
+
+-- | Similar to 'layoutHintsWithReplacement', but relies on the user to
+-- provide the ratio for each window. @aspectRatio (rx, ry) layout@ will
+-- adapt the sizes of a layout's windows according to the provided aspect
+-- ratio, and position them inside their originally assigned area
+-- according to the @rx@ and @ry@ parameters.
+-- (0, 0) places the window at the top left, (1, 0) at the top right,
+-- (0.5, 0.5) at the center, etc.
+fixedAspectRatio
+  :: (Double, Double) -> l a -> ModifiedLayout FixedAspectRatio l a
+fixedAspectRatio = ModifiedLayout . FixedAspectRatio mempty
+
+data FixedAspectRatio a = FixedAspectRatio (M.Map Window Rational)
+                                           (Double, Double)
+  deriving (Read, Show)
+
+instance LayoutModifier FixedAspectRatio Window where
+  -- | Note: this resembles redoLayout from "XMonad.Layout.LayoutHints".
+  -- The only difference is relying on user defined aspect ratios, and
+  -- using the 'adj' function defined below instead of 'mkAdjust'
+  pureModifier (FixedAspectRatio ratios placement) _ (Just s) xs =
+    (xs', Nothing)
+   where
+    xs' =
+      map (\x@(_, r) -> second (placeRectangle placement r) $ applyHint x) xs
+    applyHint (win, r@(Rectangle x y w h)) =
+      let ar       = M.lookup win ratios
+          (w', h') = maybe (w, h) (adj (w, h)) ar
+      in  (win, if isInStack s win then Rectangle x y w' h' else r)
+
+  pureModifier _ _ _ xs = (xs, Nothing)
+
+  handleMess (FixedAspectRatio ratios placement) mess
+    | Just DestroyWindowEvent { ev_window = w } <- fromMessage mess
+    = return . Just $ FixedAspectRatio (deleted w) placement
```

Comment:
I'm testing this with https://github.com/xmonad/xmonad/pull/284, and it doesn't seem to work. Is there somewhere else I also need to handle `DistroyWindowEvent`?

Reply:
If you're using NewSelect from LayoutCombinators, then it needs to be added there. Or use https://github.com/xmonad/xmonad/pull/281

Reply:
That's the thing, I'm not using any of those... I'll need to debug what's the problem here

Reply:
alright, https://github.com/xmonad/xmonad/pull/284/commits/eda5dfbe288c4a0a0174dbfd65027386cc38266f fixes it